### PR TITLE
fix exception type for sparse LDLᵀ

### LINKED ISSUE #667: ldlt factorization throws a PosDefException
https://github.com/JuliaLang/julia/blob/master/stdlib/SuiteSparse/src/cholmod.jl#L1449 gives an inaccurate error message since the input need not be SPD

### LINKED ISSUE #33351: ldlt factorization throws a PosDefException
https://github.com/JuliaLang/julia/blob/master/stdlib/SuiteSparse/src/cholmod.jl#L1449 gives an inaccurate error message since the input need not be SPD

### PR DESCRIPTION
**Pull request by stevengj**
Fixes JuliaLang/LinearAlgebra.jl#667.

### CODE REVIEW

**Comment by andreasnoack** [comment]
Unfortunately, the fix here will probably be a bit more complicated. The current fix is still not accurate as `sparse([0 1; 1 1])` is not singular but will still fail because of the zero pivot. I suppose the check in https://github.com/JuliaLang/julia/blob/4f17558e4c71908e8ba40d24e37bdacd5e2f2e5e/stdlib/SuiteSparse/src/cholmod.jl#L745-L750 was meant to handle this but apparently it doesn't
```julia
julia> cholesky(sparse([0 1; 1 1]), check=false)\ones(2)
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
 [1] solve(::Int32, ::SuiteSparse.CHOLMOD.Factor{Float64}, ::SuiteSparse.CHOLMOD.Dense{Float64}) at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:747
 [2] \ at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:1678 [inlined]
 [3] \(::SuiteSparse.CHOLMOD.Factor{Float64}, ::Array{Float64,1}) at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:1683
 [4] top-level scope at REPL[256]:1
```

**Comment by stevengj** [comment]
> `cholesky(sparse([0 1; 1 1])`

The matrix `[0 1; 1 1]` is not positive definite, so the `PosDefException` is correct in that case.  If you use the `ldlt` function it seems to work fine:

```jl
julia> ldlt(sparse([0 1; 1 1]), check=false) \ ones(2)
ERROR: ArgumentError: factorized matrix has one or more zero pivots. Try using `lu` instead.
```

**Comment by andreasnoack** [comment]
Let me try again since I clearly wrote the last comment too early in the morning. What I think would still be wrong with this PR is
```julia
ldlt(sparse([0 1; 1 1])
```
which will now throw `SingularException` but should instead error the same way as
```julia
julia> ldlt(sparse([0 1; 1 1]), check=false)\ones(2)
ERROR: ArgumentError: factorized matrix has one or more zero pivots. Try using `lu` instead.
Stacktrace:
 [1] solve(::Int32, ::SuiteSparse.CHOLMOD.Factor{Float64}, ::SuiteSparse.CHOLMOD.Dense{Float64}) at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:749
 [2] \ at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:1678 [inlined]
 [3] \(::SuiteSparse.CHOLMOD.Factor{Float64}, ::Array{Float64,1}) at /Users/andreasnoack/julia/usr/share/julia/stdlib/v1.4/SuiteSparse/src/cholmod.jl:1683
 [4] top-level scope at REPL[283]:1
```

**Comment by stevengj** [comment]
Does SuiteSparse allow you to distinguish between the actually singular case and cases where there are just zero pivots for LDLt factorization?  If not, maybe we need to define a new `ZeroPivot` exception for `ldlt`.

**Comment by andreasnoack** [comment]
> Does SuiteSparse allow you to distinguish between the actually singular case and cases where there are just zero pivots for LDLt factorization?

I don't think it has any idea if the matrix is singular or not. It only knows that it has hit a zero pivot.

> Maybe we need to define a new `ZeroPivot` exception for `ldlt`.

Yes. That might be a good idea. Could also be useful in LU without pivoting.

**Comment by stevengj** [comment]
By the way, all of the `unsafe_load(pointer(A))` code in this file scares me a bit.  Shouldn't this be sprinkled with `GC.@preserve` calls?

**Comment by stevengj** [comment]
Added the `ZeroPivotException` and used it for both `ldlt` and non-pivoted LU.

**Comment by StefanKarpinski** [comment]
What's the status here? Does this look good to you now, @andreasnoack?
